File: packages/provide-terminal/src/provide/terminal/control/plane/bootstrap.py
from __future__ import annotations

from typing import Protocol

from provide.terminal.control.plane.capability import EngineCapabilities
from provide.terminal.control.plane.transaction import Transaction
from provide.terminal.control.plane.types import ControlPlaneConfig


class ControlPlane(Protocol):
    capabilities: EngineCapabilities

    async def open(self) -> None: ...

    async def close(self) -> None: ...

    async def migrate(self) -> None: ...

    async def begin(self) -> Transaction: ...


async def bootstrap_control_plane(config: ControlPlaneConfig) -> ControlPlane:
    if config.backend == "memory":
        from provide.terminal.control.plane.memory import MemoryControlPlane

        return MemoryControlPlane(config)
    if config.backend == "sqlite":
        from provide.terminal.control.plane.sqlite import SqliteControlPlane

        return SqliteControlPlane(config)
    raise ValueError(f"unsupported control-plane backend: {config.backend}")

File: packages/provide-uterm-csharp/src/Provide.Uterm/ControlPlane/Bootstrap.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.ControlPlane;

/// <summary>Raised for an unrecognised control-plane backend. Mirrors the
/// Python ValueError("unsupported control-plane backend: ...").</summary>
public sealed class ControlPlaneConfigurationException(string message) : Exception(message);

/// <summary>
/// Selects a control-plane engine for the configured backend. Port of
/// provide.uterm.control.plane.bootstrap.bootstrap_control_plane and the Go
/// controlplane/bootstrap package.
/// </summary>
public static class Bootstrap
{
    /// <summary>
    /// Constructs an engine for <paramref name="backend"/>. The caller owns the
    /// lifecycle: open, migrate, and dispose.
    /// </summary>
    public static IEngine New(string? backend, string? databaseUrl)
    {
        var selected = string.IsNullOrWhiteSpace(backend) ? "memory" : backend.Trim().ToLowerInvariant();
        return selected switch
        {
            "memory" => new MemoryEngine(),
            "sqlite" => new SqliteEngine(
                string.IsNullOrWhiteSpace(databaseUrl) ? ":memory:" : databaseUrl),
            _ => throw new ControlPlaneConfigurationException(
                "unsupported control-plane backend: " + backend),
        };
    }

    /// <summary>
    /// Constructs an engine and readies it for use — open, then migrate. Migrate
    /// matters: nothing read a store before the graphical-target registry, so a
    /// missing schema would surface only at first use.
    /// </summary>
    public static async Task<IEngine> OpenAsync(
        string? backend, string? databaseUrl, CancellationToken ct = default)
    {
        var engine = New(backend, databaseUrl);
        try
        {
            await engine.OpenAsync(ct).ConfigureAwait(false);
            await engine.MigrateAsync(ct).ConfigureAwait(false);
            return engine;
        }
        catch
        {
            await engine.CloseAsync(CancellationToken.None).ConfigureAwait(false);
            throw;
        }
    }
}
